### PR TITLE
ARROW-7741: [C++] Add a enum to control parquet engine version

### PR DESCRIPTION
**Pull request by emkornfield**
There are going to be code paths added to read/write nested data.
Until these are fully validated we should have a flag to switch
the code on/off (new code to be implemented in future PRs)

### CODE REVIEW

File: cpp/src/parquet/properties.h [hunk 1]
```diff
@@ -535,6 +535,19 @@ static constexpr bool kArrowDefaultUseThreads = false;
...
+// of code to read/write data to Parquet.  This enum allows
+// selection of the given code.
+enum class ArrowEngineVersion {
+  // The current default engine. This doesn't support mixed
+  // struct list date.
+  kDefault = 0,
+
+  // Not fully implemented.  When done this will support reading/
+  // writing all forms of nested data.
+  kNextGen = 1
```
Comment (pitrou) [inline cpp/src/parquet/properties.h]: Will number 2 be named `kNextNextGen`? :-) I suggest we find other names (perhaps simply version names: `v1`, `v2`).

(also, I don't think the `k` prefix is necessary if it's a qualified enum member?)

**Review by pitrou — COMMENTED** [review]
Will there be a benefit to keeping the old engine version? Will the new version break compatibility or reduce performance somehow?

File: cpp/src/parquet/properties.h (hunk 1 above)
Comment (emkornfield) [reply cpp/src/parquet/properties.h]: I think qualified enums are supposed to be treated like [constants](https://google.github.io/styleguide/cppguide.html#Enumerator_Names).  In this case I changed to V1 and V2

**Comment by emkornfield** [comment]
> Will there be a benefit to keeping the old engine version? Will the new version break compatibility or reduce performance somehow?

I'm hoping at some point the two can be merged but would like to keep them initially so that if there is any performance regression or bugs users can have a path to revert to old behavior (ideally if I am careful without rolling back CLs)

**Comment by wesm** [comment]
It seems like we should merge the Parquet nested patches into a feature branch and keep iterating on it until the performance regressions are gone, instead of maintaining multiple code paths. Thoughts?

This patch seems also to revert some submodule changes

**Comment by emkornfield** [comment]
Per mailing list reply, I'd prefer to avoid a feature branch if possible.  I'll close this PR for now and try to make incremental changes that don't have performance regressions and then we can determine if the final integrations should be hidden by a flag.

**Pull request closed without merging**